File: ui/packages/ui/src/search/engine/score.ts
import type { SearchField, SearchHitKind } from "./types";

export const SEARCH_KIND_SCORE: Record<SearchHitKind, number> = {
  exact: 100,
  prefix: 40,
  contains: 10,
};

export const DEFAULT_SEARCH_FIELD_WEIGHT = 1;

export function kindScore(kind: SearchHitKind): number {
  return SEARCH_KIND_SCORE[kind];
}

export function fieldWeight(field: SearchField): number {
  const weight = field.weight;
  if (typeof weight !== "number" || !Number.isFinite(weight)) return DEFAULT_SEARCH_FIELD_WEIGHT;
  return weight > 0 ? weight : 0;
}

export function scoreKind(kind: SearchHitKind, weight: number): number {
  if (weight <= 0) return 0;
  return kindScore(kind) * weight;
}

File: core/yohu-search/src/score.rs
//! 命中种类分与字段权重。

use crate::types::{SearchField, SearchHitKind};

pub const SEARCH_KIND_SCORE_EXACT: i32 = 100;
pub const SEARCH_KIND_SCORE_PREFIX: i32 = 40;
pub const SEARCH_KIND_SCORE_CONTAINS: i32 = 10;
pub const DEFAULT_SEARCH_FIELD_WEIGHT: i32 = 1;

pub fn kind_score(kind: SearchHitKind) -> i32 {
    match kind {
        SearchHitKind::Exact => SEARCH_KIND_SCORE_EXACT,
        SearchHitKind::Prefix => SEARCH_KIND_SCORE_PREFIX,
        SearchHitKind::Contains => SEARCH_KIND_SCORE_CONTAINS,
    }
}

pub fn field_weight(field: &SearchField) -> i32 {
    match field.weight {
        Some(w) if w > 0 => w,
        Some(_) => 0,
        None => DEFAULT_SEARCH_FIELD_WEIGHT,
    }
}
